Add QOL changes to permission middleware

BowsiePup opened the pull request:
Changes:
- Add human readable permission names to error message
- If embed perm is missing response will be in plain text even with rose error logging enabled
- Only show the perms that are actually missing on user and bot
- Add option for rose error logging
- Put each required perm in an inline embed

BowsiePup committed:
Update.
Changes:
-  Add human readable to error message
- If embed perm is missing response will be in plain text
- Only show the perms that are actually missing on user and bot
- Add option for rose error logging

jpbberry commented on index.js:
```diff
@@ -1,15 +1,52 @@
 module.exports = ({
-  user = (ctx) => `You're the following permissions: ${ctx.command.userPerms.join(', ')}`,
-  my = (ctx) => `I'm missing the following permissions: ${ctx.command.myPerms.join(', ')}`
+  humanReadable = {
```
Why is this an object passed to the method? Just make it a constant

BowsiePup replied:
language support??!

jpbberry replied:
alright

jpbberry replied:
You should still make the default a constant, and possibly export it too, and just make the humanReadable = the constant.

alexanderpaolini replied:
For better language support, wouldn't you want to be able to set `humanReadable` to a function as well? This would mean that you could have different languages for different guilds without rewriting the middleware to better suit the bot.

jpbberry commented on index.js:
```diff
@@ -1,15 +1,52 @@
+  },
+  my = (ctx) => `I am missing the following permissions: \`${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => humanReadable[p]).join('`, `')}\``,
+  user = (ctx) => `You are missing the following permissions: \`${ctx.command.userPerms.filter(p => !ctx.userPerms(p)).map(p => humanReadable[p]).join('`, `')}\``,
+  sendAsRoseError = false
```
Shouldn't default to false because this will change what the previous behavior was

jpbberry commented on index.js:
```diff
@@ -1,15 +1,52 @@
+      if (ctx.command.hasOwnProperty('myPerms') && !ctx.command.myPerms.every(x => ctx.myPerms(x))) {
+          if (sendAsRoseError && ctx.myPerms('embed')) throw new Error(my(ctx))
+          ctx.reply(my(ctx))
+          return false
```
Default to ctx.send or make it optional since deleted messages can't be replied to.

jpbberry commented on types.d.ts:
```diff
@@ -17,8 +17,10 @@ declare module 'discord-rose/dist/typings/lib' {
 type msgFunction = (ctx: CommandContext) => string
 
 declare const _default: (msgs?: {
+  humanReadable: object,
```
This is also not correct typings

jpbberry commented on types.d.ts:
```diff
@@ -17,8 +17,10 @@ declare module 'discord-rose/dist/typings/lib' {
+  sendAsRoseError: boolean
 }) => (ctx: CommandContext) => boolean
 
 export default _default
```
Export human readable from the permissions middleware maybe.

jpbberry commented on types.d.ts:
```diff
@@ -17,8 +17,10 @@ declare module 'discord-rose/dist/typings/lib' {
 type msgFunction = (ctx: CommandContext) => string
 
 declare const _default: (msgs?: {
```
This would also no longer be categorized as `msgs`
```suggestion
declare const _default: (opts?: {
```

jpbberry commented on index.js:
```diff
@@ -1,15 +1,52 @@
+      webhooks: 'Manage Webhooks',
+      emojis: 'Manage Emojis'
+  },
+  my = (ctx) => `I am missing the following permissions: \`${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => humanReadable[p]).join('`, `')}\``,
```
```suggestion
  my = (ctx) => `I am missing the following permissions: \`${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => humanReadable[p] || p).join('`, `')}\``,
```
Also should do this in case of a new permission taking a bit to add.

BowsiePup committed:
commit fixing requested changes

BowsiePup committed:
update the default custom messages in readme

BowsiePup committed:
add more info about the default messages

BowsiePup commented:
Dumb bot doesnt alert for re-review request of new commits. smh.

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,57 @@
+const exportFunc = module.exports = ({
+  humanReadable = humanReadable,
+  my = (ctx) => `I am missing the following permissions: \`${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) || p).join('`, `')}\``,
+  user = (ctx) => `You are missing the following permissions: \`${ctx.command.userPerms.filter(p => !ctx.userPerms(p)).map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) || p).join('`, `')}\``,
```
`??` should probably be used rather than `||` as the OR operator will replace an empty string with the permissions name which could be an unintended behavior

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,57 @@
+          return false
+      }
+      if (ctx.command.hasOwnProperty('userPerms') && !ctx.command.userPerms.every(x => ctx.hasPerms(x))) {
+          if (sendAsRoseError) throw new Error(user(ctx))
```
Here you do not check for `ctx.myPerms('embed')` yet you do check on line 44, perhaps one should be switched to the other. Since it is an option and throwing an error does not necessarily mean that an embed will be used, the check is probably unnecessary,

BowsiePup replied:
that throw new error is handled by discord rose and its error handle only sends embeds [here](https://github.com/discord-rose/discord-rose/blob/acaf79ae5d4644b869371ab37e2e8a03e1615bab/src/structures/CommandHandler.ts#L45)

alexanderpaolini replied:
That function should probably be changed to check for embed permissions before sending an embed. Personally I would move the check to the error handling function rather than keeping it here since developers can configure the function to handle it without an embed which would cause issues if the bot does not have the embed permissions. @jpbberry, this is up to you.

BowsiePup replied:
for now i will keep the checks in place... if berry decides to have the default handler check for embed issues then it can be removed

jpbberry replied:
I think it shouldn't override the error function and should just throw even if a missing embed permissions occurs. @Million900o feel free to open a PR on the main discord-rose package to change the default error function to properly handle embed errors.

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,57 @@
-module.exports = ({
-  user = (ctx) => `You're the following permissions: ${ctx.command.userPerms.join(', ')}`,
-  my = (ctx) => `I'm missing the following permissions: ${ctx.command.myPerms.join(', ')}`
+const humanReadable = {
```
Does doing separating the variables work for you? I was not able to get it to work. Starting the cluster would throw the error: `ReferenceError [Error]: Cannot access 'humanReadable' before initialization`

alexanderpaolini replied:
The error was thrown on line 36.

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,57 @@
+
+const exportFunc = module.exports = ({
+  humanReadable = humanReadable,
+  my = (ctx) => `I am missing the following permissions: \`${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) || p).join('`, `')}\``,
```
This default response does not work well with the default error handling embed. It leaves a trailing backtick on the next line. Adding `\n` to the end fixes the issue, however, it might be better to just forgo the backticks.

jpbberry commented on types.d.ts:
```diff
@@ -1,24 +1,86 @@
   }
 }
 
+type humanReadableBits = {
```
```suggestion
interface humanReadableBits {
  [key in bitKey]: string
}
```

BowsiePup replied:
Complains about that
![image](https://user-images.githubusercontent.com/34384474/117298051-e94bcc00-ae44-11eb-89c4-dc22cb8ee1a7.png)

alexanderpaolini replied:
Is there a reason to use `key in bitKey` rather than `key: keyof typeof bits`?

jpbberry replied:
Sorry it'd be
```ts
type humanReadable = {
  [key in bitKey[number]]: string
}
```

BowsiePup committed:
Changes.
- Fix types.d.ts
- Update readme with new default response
- remove embed perm checking
- change to perm map to nullness

jpbberry commented:
What. Why'd you remove the humanReadable just to paste it twice?

jpbberry commented:
Also please address millions reviews.

BowsiePup commented:
that should be everything now

BowsiePup commented:
> What. Why'd you remove the humanReadable just to paste it twice?

https://github.com/discord-rose/permissions-middleware/pull/1#discussion_r627348743

jpbberry commented:
The error is due to the fact that the constant is named the same thing as `humanReadable` options parameter. Change the constant name rather than copy and pasting

BowsiePup committed:
Commit.

- Actually push types.d.ts and readme
- Rename const cause issues

BowsiePup commented:
there

jpbberry commented on types.d.ts:
```diff
@@ -1,24 +1,56 @@
+/**
+ * A JSON of the default provided bits
+ */
+export const humanReadable: {
+  key: (keyof typeof bits)[],
+  value: string
+}
```
```suggestion
export const humanReadable: humanReadableBits
```
Read my other comment as well since this wont work with what's there right now

jpbberry commented on types.d.ts:
```diff
@@ -1,24 +1,56 @@
   }
 }
 
+type humanReadable = {
```
I think you meant to name this `humanReadableBits`

BowsiePup committed:
make humanReadableBits optional

BowsiePup committed:
change export const humanReadable

alexanderpaolini commented on README.md:
```diff
@@ -48,10 +68,13 @@ When using `permissionsMiddleware()` you can pass a custom message object with a
 {
-  user = (ctx) => `You're the following permissions: ${ctx.command.userPerms.join(', ')}`,
-  my = (ctx) => `I'm missing the following permissions: ${ctx.command.myPerms.join(', ')}`
+  my = (ctx) => `I am missing the following permissions: ${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) || p).join(', ')}`,
```
It would probably be better if these `my` and `user` passed the missing permissions as if the user has to do the check for missing permissions, there is no point to use this. Additionally, passing through this specific function as an option creates issues with the `humanReadable`. Essentially, the paramaters `humanReadable` and `my` and `user` would actually be exclusive. I suggest changing `msgFunction` to
```js
type msgFunction = (perms: (keyof typeof bits)[],  humanReadable: humanReadableBits, ctx: CommandContext) => string
```
or something similar.

BowsiePup replied:
can u elaborate

alexanderpaolini replied:
Currently, if I want to change the function to return `Bonk! I'm missing the permissions: Kick Members, Ban Members`, with the current changes I would need to copy the `humanReadable` object. There is no way to access it in the passed function. Ex. 
```js
worker.commands.middleware(permissionsMiddleware({
  my: (ctx) => `I am missing the following permissions: ${ctx.command.myPerms.filter(p => !ctx.myPerms(p)).map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) ?? p).join(', ')}`
}))
```
Throws the error: `humanReadable is not defined`. This is because the `humanReadable` would need to be in the scope of the defined function. 
By passing both the missing permissions and the `humanReadable` object, I could make it
```js
worker.commands.middleware(permissionsMiddleware({
  my: (perms, humanReadable, ctx) => `Bonk! I'm missing the permissions: ${perms.map(p => humanReadable[p] ?? p).join(', ')}`
}))
```
Which is much more readable and reduces duplicated code.

BowsiePup replied:
permissionsMiddleware.humanReadable is accessible tho.

BowsiePup replied:
but yes i see where you are coming from

jpbberry replied:
I think passing `humanReadable` to the function is unnecessary and limiting, so no. I was gonna suggest that you could make it optionally `perms` to be the readable, however I don't even like that because it's unexpected.

alexanderpaolini replied:
Oh, you were exporting it, I didn't think about that.

alexanderpaolini replied:
I think it would be more unexpected to be given a missing permissions function without the missing permissions. At the very least, it would make it easier and more readable being able to just do `perms.join(', ')` rather than filtering the command's permissions.

BowsiePup replied:
so no changes?

BowsiePup replied:
well it would still be `perms.map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) ?? p).join(', ')`

BowsiePup replied:
unless you are thinking about putting the map to human readable outside the my and user. then there would be no need for passing it into the function

jpbberry replied:
Sure, just don't pass humanReadable to the function that looks ugly.

alexanderpaolini replied:
Only for code that must account for both. A user would only be able to pass an object OR a function. It would therefore be
```js
perms.map(p => humanReadable(ctx, p) ?? p).join(', ')
```
OR
```js
perms.map(p => humanReadable[p] ?? p).join(', ')
```
A user would not need to include both

BowsiePup replied:
okay so just perms and ctx

perms will be pre human readablified before going into my and user

jpbberry replied:
No they shouldn't be, special ops can be used to recognize the missing permissions so they shouldn't be made human before being passed

alexanderpaolini reviewed (changes requested):
This does not actually make any sense in an actual use-case.

BowsiePup committed:
add perms passed into my and user, update readme

BowsiePup commented:
hopefully final revision

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,59 @@
-      throw new Error(user(ctx))
+    if (ctx.command.hasOwnProperty('myPerms') && !ctx.command.myPerms.every(x => ctx.myPerms(x))) {
+      const perms = ctx.command.myPerms.filter(p => !ctx.myPerms(p))
+      if (sendAsRoseError) throw new Error(my(perms, ctx))
```
It might be better to use `return ctx.error()` rather than `throw new Error()`, but that is probably a useless change that has no real effect. Also, if the bot is missing the `sendMessages` permission or the replied to message was deleted, an error would be thrown when the response is sent. This realistically is not an issue but it might be better to catch it regardless. This is up to @jpbberry if he wants this to be changed as it should still be handled by the command handler.

jpbberry replied:
I'll make the transition to ctx.error() with all the middlewares together, not right now though.

jpbberry commented on index.js:
```diff
@@ -1,15 +1,59 @@
+  humanReadable = humanReadableStrings,
+  my = (perms, ctx) => `I am missing the following permissions: ${perms.map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) ?? p).join(', ')}`,
+  user = (perms, ctx) => `You are missing the following permissions: ${perms.map(p => (typeof humanReadable === 'function' ? humanReadable(ctx, p) : humanReadable[p]) ?? p).join(', ')}`,
+  sendAsRoseError = true,
+  makeResponseAReply = false
```
After some thought I want you to remove these extra options, I find them useless and if anything more confusing then anything else. Instead only work with middleware `throw new Error()` this still leaves plenty of room for users to handle the errors normally via customizing their errors on the CommandHandler.

jpbberry commented:
I also realize something about the ability to customize humanReadable is fundamentally flawed, because if you're going to customize the language, then the my() and user() functions would need to change anyways, which would then disregard the ability to customize humanReadable because it's only applied to the default functions.

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,59 @@
-    if (ctx.command.myPerms && !ctx.command.myPerms.every(x => ctx.myPerms(x))) {
-      throw new Error(my(ctx))
+    if (ctx.command.hasOwnProperty('userPerms') && !ctx.command.userPerms.every(x => ctx.hasPerms(x))) {
+      const perms = ctx.command.userPerms.filter(p => !ctx.userPerms(p))
```
`ctx.userPerms` should be replaced with `ctx.hasPerms`

BowsiePup commented:
> I also realize something about the ability to customize humanReadable is fundamentally flawed, because if you're going to customize the language, then the my() and user() functions would need to change anyways, which would then disregard the ability to customize humanReadable because it's only applied to the default functions.

I think leaving it in would be good. What if you want to change the format of the error but still be in english?

jpbberry commented:
Like I said, that's kinda fundamentally flawed. And if they do want that they can just define their own user() and my() function.

alexanderpaolini commented:
The `my` and `user` functions should also be able to return promises.

BowsiePup committed:
Changes.

- Remove sendAsRoseError & makeResponseAReply
- Move error response to return ctx.error()
- change userPerm checking to hasPerms
- Remove humanReadable option
- Change the default responses to not check if humanReadable is a function

jpbberry commented on README.md:
```diff
@@ -35,29 +35,98 @@ worker.commands
 {
-  user = (ctx) => `You're the following permissions: ${ctx.command.userPerms.join(', ')}`,
-  my = (ctx) => `I'm missing the following permissions: ${ctx.command.myPerms.join(', ')}`
+  my = (perms, _ctx) => `I am missing the following permissions: ${perms.map(p => humanReadable[p] ?? p).join(', ')}`,
+  user = (perms, _ctx) => `You are missing the following permissions: ${perms.map(p => humanReadable[p] ?? p).join(', ')}`,
```
perms and _ctx are the wrong way around, also change `_ctx` to ctx for both the example and code please.

jpbberry commented on README.md:
```diff
@@ -35,29 +35,98 @@ worker.commands
 {
-  user: (ctx) => ...,//
-  my: (ctx) => ... //
+  user: (perms, ctx) => ...,//
+  my: (perms, ctx) => ... //
```
Same thing here

jpbberry commented on README.md:
````diff
@@ -35,29 +35,98 @@ worker.commands
+Creating custom messages but using the provided readable permissions:
+```js
+permissionsMiddleware({
+  user: (perms, ctx) => `You don't have the required permissions you need: ${ctx.command.userPerms.map(p => permissionsMiddleware.humanReadable[p])}`,
+  my: (perms, ctx) => `You don't have the required permissions you need: ${ctx.command.myPerms.map(p => permissionsMiddleware.humanReadable[p])}`
````
same thing here

jpbberry commented on types.d.ts:
```diff
@@ -1,24 +1,41 @@
+  [key in bitKey[number]]?: string
+}
+
+type msgFunction = (perms: (keyof typeof bits)[], ctx: CommandContext) => string
```
same thing here, and by the way no don't just change the order in the code, ctx has to come first otherwise it'd be a breaking change.

BowsiePup committed:
make my and user accept promises.

jpbberry commented on README.md:
````diff
@@ -35,29 +35,98 @@ worker.commands
+
+## Default readable permissions
+
+```json
+{
+  createInvites: 'Create Invites',
+  kick: 'Kick Members',
+  ban: 'Ban Members',
+  administrator: 'Administrator',
+  manageChannels: 'Manage Channels',
+  manageGuild: 'Manage Server',
+  addReactions: 'Add Reactions',
+  auditLog: 'View Audit Log',
+  prioritySpeaker: 'Priority Speaker',
+  stream: 'Stream',
+  viewChannel: 'View Channel(s)',
+  sendMessages: 'Send Messages',
+  tts: 'Send Text-to-Speech Messages',
+  manageMessages: 'Manage Messages',
+  embed: 'Embed Links',
+  files: 'Attach Files',
+  readHistory: 'Read Message History',
+  mentionEveryone: 'Mention \@everyone, \@here, and All Roles',
+  externalEmojis: 'Use External Emoji',
+  viewInsights: 'View Server Invites',
+  connect: 'Connect (Voice)',
+  speak: 'Speak (Voice)',
+  mute: 'Mute (Voice)',
+  deafen: 'Deafen (Voice)',
+  move: 'Move (Voice)',
+  useVoiceActivity: 'Use Voice Activity',
+  nickname: 'Change Nickname',
+  manageNicknames: 'Manage Nicknames',
+  manageRoles: 'Manage Roles',
+  webhooks: 'Manage Webhooks',
+  emojis: 'Manage Emojis'
+}
````
I find this useless to be in the README. Especially since it can change and it's not static/breaking.

BowsiePup commented:
Changes.
- Remove sendAsRoseError & makeResponseAReply
- Move error response to return ctx.error()
- change userPerm checking to hasPerms
- Remove humanReadable option
- Change the default responses to not check if humanReadable is a function
- Make my and user accept promises.

BowsiePup committed:
Reorder params, update readme

BowsiePup committed:
fix relative link in readme

BowsiePup commented:
Reorder params, update readme
fix relative link in readme

jpbberry commented:
I unresolved an issue you didn't address

alexanderpaolini commented on types.d.ts:
```diff
@@ -1,24 +1,41 @@
+  [key in bitKey[number]]?: string
+}
+
+type msgFunction = (ctx: CommandContext, perms: (keyof typeof bits)[]) => string
```
`msgFunction` should return `Promise<string>` as well as `string`

alexanderpaolini commented on types.d.ts:
```diff
@@ -1,24 +1,41 @@
 import { bits } from 'discord-rose/dist/utils/Permissions'
 import { CommandContext } from "discord-rose/dist/structures/CommandContext";
```
`discord-rose/dist/structures/CommandContext` should be replaced with `discord-rose/dist/typings/lib`

BowsiePup committed:
remove _ in the read me for ctx

BowsiePup committed:
update typings

alexanderpaolini commented on index.js:
```diff
@@ -1,15 +1,50 @@
-    if (ctx.command.myPerms && !ctx.command.myPerms.every(x => ctx.myPerms(x))) {
-      throw new Error(my(ctx))
+    if (ctx.command.hasOwnProperty('userPerms') && !ctx.command.userPerms.every(x => ctx.hasPerms(x))) {
+      return ctx.error(await user(ctx, ctx.command.hasPerms.filter(p => !ctx.userPerms(p))))
```
`ctx.hasPerms` is the function to check for permissons. `ctx.userPerms` is the array of permissions. They need to be swapped here.

alexanderpaolini commented on types.d.ts:
```diff
@@ -1,24 +1,41 @@
 import { bits } from 'discord-rose/dist/utils/Permissions'
-import { CommandContext } from "discord-rose/dist/structures/CommandContext";
+import { CommandContext } from "discord-rose/dist/typings/lib";
```
To stay consistent with formatting, the double quotes should be swapped with single quotes and the semicolon should be removed. This was not caused by the changes made in this pull request but since you are already changing things, you might as well just change the formatting.
